Create copilot-instructions.md (#115917)
dotnet/runtime: .github/copilot-instructions.md

Important: **Ensure the code compiles and the tests pass.** Use the "Building & Testing in dotnet/runtime" instructions below.

Additionally,

> * Don't leave trailling whitespaces.

Author: > Comments suppressed due to low confidence (1)
> **.github/copilot-instructions.md:5**
> 
> * There is a typo in the word 'trailling'; it should be 'trailing'.
> 
> ```
> * Don't leave trailling whitespaces.
> ```

```suggestion
* Don't leave trailing whitespaces.
```

- **Mono Runtime:** Changes in `src/mono/`
- **Libraries:** Changes in `src/libraries/`
- **WASM/WASI Libraries:** Changes in `src/libraries/` *and* the affected library targets WASM or WASI *and* the changes are included for the target (see below for details).
- If none above apply, it is most possibly an infra-only or a docs-only change. Skip build and test steps.

**WASM/WASI Library Change Detection**
A change is considered WASM/WASI-relevant if:
- The relevant `.csproj` contains explicit Browser/WASM or WASI targets (look for `<TargetFrameworks>`, `$(TargetPlatformIdentifier)`, or `Condition` attributes referencing `browser` or `wasi`, as well as `TARGET_BROWSER` or `TARGET_WASI` constants), **and**
- The changed file is not excluded from the build for that platform in any way with a `Condition` attribute on `<ItemGroup>` or `<Compile>`.

### 1.2. Baseline Setup

Ensure you have a full successful build of the needed runtime+libraries as a baseline.

1. Checkout `main` branch

2. From the repository root, run the build depending on the affected component. If multiple components are affected, subsequently run and verify the builds for all of them.

>    - **CoreCLR (CLR):** `./build.sh clr+libs`

Reviewer: ```suggestion
   - **CoreCLR (CLR):** `./build.sh clr+libs+host`
```

The host may or may not be necessary depending on what's going to be run, but this is better safe than sorry.

Author: I didn't notice it mentioned in the docs (within a "typical scenario") anywhere beside running the Mono runtime tests 😅
Can it be more applicable to runtime, but not that much for the libraries?

I'm just a bit reluctant to add something which is not described the workflow docs as typical and might not be necessary. But I might be missing something, I'm not an expert on building anything beside libraries 😅

Author: I mean, I've seen that the host exists, but when I open the [building docs](https://github.com/dotnet/runtime/blob/main/docs/workflow/building/coreclr/README.md), that's what I see:

**The Basics**

... main workflow ... build the CoreCLR ...

```
./build.sh -subset clr <other args go here>
```

> _// where are these other args?_

**Build Results**

corerun: ...
coreclr: ...
System.Private.CoreLib.dll: ...

> _// uh-huh_

**What to do with the Build**

> _// uh-huh_

**The Core_Root for Testing Your Build**

```
./build.sh -subset clr+libs -runtimeConfiguration Debug -librariesConfiguration Release
```

> _// cool seems legit_

```
./src/tests/build.sh -x64 -checked -generatelayoutonly
```

> _// ok these are tests already_

**The Dev Shipping Packs**

... full runtime NuGet packages and installer ... more production-esque scenario...

> _// naaah, I don't need that yet_

😅

   - **Mono Runtime:** `./build.sh mono+libs`

>    - **Libraries:** `./build.sh clr+libs -rc release`

Reviewer: ```suggestion
   - **Libraries:** `./build.sh clr+libs+host -rc release`
```

   - **WASM/WASI Libraries:** `./build.sh mono+libs -os browser`

3. Verify the build completed without error.
   - _If the build failed, report the failure and don't proceed with the changes._
4. From the repository root:
   - Configure PATH: `export PATH="$(pwd)/.dotnet:$PATH"`
   - Verify SDK Version: `dotnet --version` should match `sdk.version` in `global.json`.
5. Switch back to the working branch.

---

## 2. Iterative Build and Test Strategy

1. Apply the intended changes

> 2. **Attempt Build.** If the build fails, attempt to fix and retry the step (up to 5 attempts).

Reviewer: This may work sometimes, but might fail if it tries to build the "wrong" thing. For instance, if it makes a change to the CLR but doesn't rebuild the CLR partition. The libraries build also does deployment, so certain actions may require you to basically rebuild everything.

Author: > if it makes a change to the CLR but doesn't rebuild the CLR partition

technically this should be covered by the instructions above.
and the agent also does a decent job of trying out different approaches on how to build stuff.

This part is more about actual build failures (like trailing whitespaces 🙃 or, say, the library is targeting netstandard so some of the changes might try to use APIs that are not available)

And in case it goes completely off track, it is asked to limit the number of attempts.

> The libraries build also does deployment, so certain actions may require you to basically rebuild everything.

I don't think I ever needed to do that 🤔 this sounds a bit like a corner case... (or maybe it's just networking libs that aren't affected?)

But in general, I think we don't have to (maybe even _shouldn't_) cover every single possibility here. If we do, we will end up including all the existing workflow docs, and add some more on top. And I'm not sure the result will be helpful for the 90% usecase.

      - If the problem is in the test, attempt to fix and retry the step (up to 5 attempts).
      - If the problem is in the source, reconsider the full changeset, attempt to fix and repeat the workflow.
4. **Workflow Iteration:**
   * Repeat build and test up to 5 cycles.
   * If issues persist after 5 workflow cycles, report failure.
   * If the same error persists after each fix attempt, do not repeat the same fix. Instead, escalate or report with full logs.

When retrying, attempt different fixes and adjust based on the build/test results.

### Success Criteria

* **Build:**
   - Completes without errors.
   - Any non-zero exit code from build commands is considered a failure.
* **Tests:**
   - All tests must pass (zero failures).
   - Any non-zero exit code from test commands is considered a failure.

* **On success:** Report completion;
  - Otherwise, report error(s) with logs for diagnostics.
    - Collect logs from `artifacts/log/` and the console output for both build and test steps.
    - Attach relevant log files or error snippets when reporting failures.

---

## 3. CoreCLR (CLR) Workflow

From the repository root:

- Build: `./build.sh -subset clr`

> - Run tests: `cd src/tests && ./build.sh`

Reviewer: This builds the tests. You need to run `run.sh` to run them. It's mentioned at the bottom of the link below.

Author: Thanks

### 5.1. How To: Identify Affected Libraries

For each changed file under `src/libraries/`, find the matching library and its test project(s).
Most libraries use:

- Source: `src/libraries/<LibraryName>/src/<LibraryName>.csproj`

- Tests (single):
    - `src/libraries/<LibraryName>/tests/<LibraryName>.Tests.csproj`
    - OR `src/libraries/<LibraryName>/tests/<LibraryName>.Tests/<LibraryName>.Tests.csproj`

- Tests (multiple types):
    - `src/libraries/<LibraryName>/tests/FunctionalTests/<LibraryName>.Functional.Tests.csproj`
    - `src/libraries/<LibraryName>/tests/UnitTests/<LibraryName>.Unit.Tests.csproj`
    - Or similar.

---

### 5.2. How To: Build and Test Specific Library

If only one library is affected:

1. **Navigate to the library directory:**
   `cd src/libraries/<LibraryName>`

2. **Build the library:**
   `dotnet build`

3. **Build and run all test projects:**

    - For each discovered `*.Tests.csproj` in the `tests` subdirectory:
      `dotnet build /t:test ./tests/<TestProject>.csproj`

        - *Adjust path as needed. If in doubt, search with `find tests -name '*.csproj'`.*

>     - `dotnet build /t:test` is generally preferred over `dotnet test`

Reviewer 2: This was fixed last year, but I checked and it does appear to have crept back in  https://github.com/dotnet/runtime/issues/94183

---

## 6. WebAssembly (WASM) Libraries Workflow

From the repository root:

- Build:
  `./build.sh libs -os browser`

- Run tests:
  `./build.sh libs.tests -test -os browser`

- More info:
    - [Build libraries for WebAssembly](https://github.com/dotnet/runtime/blob/main/docs/workflow/building/libraries/webassembly-instructions.md)
    - [Testing Libraries on WebAssembly](https://github.com/dotnet/runtime/blob/main/docs/workflow/testing/libraries/testing-wasm.md)

---

## 7. Additional Notes

### 7.1. Troubleshooting

- **Shared Framework Missing**

>     - If the build fails with an error "The shared framework must be built before the local targeting pack can be consumed.", build the runtime (clr or mono) first.
>       E.g., from the repo root, run `./build.sh clr -rc release`.

Author: ```suggestion
    - If the build fails with an error "The shared framework must be built before the local targeting pack can be consumed.", build both the runtime (clr or mono) and the libs.
      E.g., from the repo root, run `./build.sh clr+libs -rc release` if working on Libraries on CoreCLR. Refer to the section `1.2. Baseline Setup`.
```

- **Testhost Is Missing**

    - If a test run fails with errors indicating a missing testhost, such as:
        - "Failed to launch testhost with error: System.IO.FileNotFoundException", or
        - "artifacts/bin/testhost/... No such file or directory",
      this means the required runtime was not built.

    - To resolve, build the appropriate runtime before running tests.

>       For example, run `./build.sh clr -rc release` if testing libraries.

Reviewer 2: These don't look right to me.  Building the testhost is both clr and libs, in particular https://github.com/dotnet/runtime/blob/main/src/libraries/pretest.proj

Maybe we just remove this now that you're pre-building them?

Author: The case for this failure was when the agent used `./build.sh libs -rc release`, i.e. skipped `clr` from `clr+libs`. That's why I assumed that bringing the missing part should help. But if the `a+b` combination actually brings something additional over the two things `a` and `b` done separately, then yeah, it wouldn't work, but I must say it's a bit counter-intuitive 😅

I worry a bit that removing it completely will bring problems with a similar case but on a mono runtime, but if it does, we can address it separately.
